packages/remix-react/components.tsx

```diff
@@ -1000,13 +1002,25 @@ export function useMatches(): UIMatch[] {
   return useMatchesRR() as UIMatch[];
 }
 
+// Deserialize unless it's a clientLoader/clientAction returning a non-Response
+// prettier-ignore
+type InferDataType<T = AppData> = (
+  T extends (...args: any[]) => unknown
+    ? ReturnType<T> extends Promise<Response>
+      ? SerializeFrom<T>
+      : Parameters<T> extends [ClientLoaderFunctionArgs | ClientActionFunctionArgs]
+        ? Awaited<ReturnType<T>>
+        : SerializeFrom<T>
+    : SerializeFrom<T>
+);
```

Comment:
We return `Awaited<ReturnType<T>>` only when the generic is a function that doesn't return a `Response`and has parameters of `ClientLoaderFunctionArgs` or `ClientActionFunctionArgs`